_examples/real-world-examples/server-sent-events-htmx/main.go

```diff
@@ -0,0 +1,531 @@
+`
+
+func main() {
+	var cfg config
+	err := envconfig.Process("", &cfg)
+	if err != nil {
+		panic(err)
+	}
+
+	db, err := sql.Open("postgres", cfg.DatabaseURL)
+	if err != nil {
+		panic(err)
+	}
+
+	_, err = db.Exec(migration)
+	if err != nil {
+		panic(err)
+	}
+
+	logger := watermill.NewStdLogger(false, false)
+
+	subscriber, err := googlecloud.NewSubscriber(googlecloud.SubscriberConfig{
+		ProjectID: cfg.PubSubProjectID,
+		GenerateSubscriptionName: func(topic string) string {
+			return fmt.Sprintf("%v_%v", topic, watermill.NewShortUUID())
+		},
+		SubscriptionConfig: pubsub.SubscriptionConfig{
+			ExpirationPolicy: time.Hour * 24,
+		},
+	}, logger)
+	if err != nil {
+		panic(err)
+	}
+
+	publisher, err := googlecloud.NewPublisher(googlecloud.PublisherConfig{
+		ProjectID: cfg.PubSubProjectID,
+	}, logger)
+	if err != nil {
+		panic(err)
+	}
+
+	sseRouter, err := http.NewSSERouter(http.SSERouterConfig{
+		UpstreamSubscriber: subscriber,
+		Marshaler:          http.BytesSSEMarshaler{},
+	}, logger)
+	if err != nil {
+		panic(err)
+	}
+
+	storage := NewStorage(db)
+
+	statsHandler := sseRouter.AddHandler(postStatsUpdatedTopic, statsStream{storage: storage})
+
+	e := echo.New()
+	e.Use(middleware.Recover())
+	e.Use(middleware.Logger())
+
+	e.GET("/", func(c echo.Context) error {
+		posts, err := storage.AllPosts(c.Request().Context())
+		if err != nil {
+			return err
+		}
+
+		for _, post := range posts {
+			event := PostViewed{
+				PostID: post.ID,
+			}
+
+			payload, err := json.Marshal(event)
+			if err != nil {
+				return err
+			}
+
+			msg := message.NewMessage(watermill.NewUUID(), payload)
+			err = publisher.Publish(postViewedTopic, msg)
+			if err != nil {
+				return err
+			}
+		}
+
+		var postViews []views.Post
+		for _, post := range posts {
+			postViews = append(postViews, newPostView(post))
+		}
+
+		return views.Index(postViews).Render(c.Request().Context(), c.Response())
+	})
+
+	e.POST("/posts/:id/reactions", func(c echo.Context) error {
+		postID, err := strconv.Atoi(c.Param("id"))
+		if err != nil {
+			return err
+		}
+
+		reactionID := c.FormValue("reaction_id")
+
+		var found bool
+		for _, r := range allReactions {
+			if r.ID == reactionID {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return c.String(stdHttp.StatusBadRequest, "invalid reaction ID")
+		}
+
+		event := PostReactionAdded{
+			PostID:     postID,
+			ReactionID: reactionID,
+		}
+
+		payload, err := json.Marshal(event)
+		if err != nil {
+			return err
+		}
+
+		msg := message.NewMessage(watermill.NewUUID(), payload)
+		err = publisher.Publish(postReactionAddedTopic, msg)
+		if err != nil {
+			return err
+		}
+
+		return c.NoContent(stdHttp.StatusAccepted)
+	})
+
+	e.GET("/posts/:id/stats", func(c echo.Context) error {
+		postID := c.Param("id")
+		c.Request().SetPathValue("id", postID)
+
+		statsHandler(c.Response(), c.Request())
+		return nil
+	})
+
+	router, err := message.NewRouter(message.RouterConfig{}, logger)
```

Comment:
it may be good to rename reouter to watermillRouter or something to not mess it with http